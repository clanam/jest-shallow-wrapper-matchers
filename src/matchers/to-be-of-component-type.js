import { ShallowWrapper } from 'enzyme';
import { matcherHint, printReceived } from 'jest-matcher-utils';

function passMessage(expected, actual) {
  return () =>
    matcherHint('.not.toBeOfComponentType', 'received', '') +
    `\n\nExpected element not to be of component type ${expected}:\n` +
    `  ${printReceived(actual)}`;
}

function failMessage(expected, actual) {
  return () =>
    matcherHint('.toBeOfComponentType', 'received', '') +
    `\n\nExpected element to be of component type ${expected}:\n` +
    `  ${printReceived(actual)}`;
}

/** Checks if a shallow-rendered enzyme element to be of a given component type. */
function toBeOfComponentType(receivedElem, expectedComponentType) {
  if (!(receivedElem instanceof ShallowWrapper)) {
    throw new Error(`Expected ${receivedElem} to be an enzyme ShallowWrapper!`);
  }

  const actualType = receivedElem.name();
  const pass = expectedComponentType === actualType;

  const msgFxn = pass ? passMessage : failMessage;
  const message = msgFxn(expectedComponentType, actualType);

  return { pass, message };
}

export { toBeOfComponentType };
