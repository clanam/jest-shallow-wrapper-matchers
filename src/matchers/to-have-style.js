import { ShallowWrapper } from 'enzyme';
import { matcherHint, printReceived } from 'jest-matcher-utils';
import { getStyleProperty } from '../helpers/get-style';

function print(name, value) {
  return `{${name}: ${value}}`;
}

function passMessage(styleName, expectedValue, actualValue) {
  return () =>
    matcherHint('.not.toHaveStyle', 'received', '') +
    `\n\nExpected element not to have style ${print(styleName, expectedValue)}.\n` +
    `  ${printReceived(print(styleName, actualValue))}`;
}

function failMessage(styleName, expectedValue, actualValue) {
  return () =>
    matcherHint('.toHaveStyle', 'received', '') +
    `\n\nExpected element to have style ${print(styleName, expectedValue)}.\n` +
    `  ${printReceived(print(styleName, actualValue))}`;
}

/** Checks if a shallow-rendered enzyme element has a given style. */
function toHaveStyle(receivedElem, styleName, styleValue) {
  if (!(receivedElem instanceof ShallowWrapper)) {
    throw new Error(`Expected ${receivedElem} to be an enzyme ShallowWrapper!`);
  }

  const actualStyleValue = getStyleProperty(receivedElem, styleName);
  const pass = actualStyleValue === styleValue;

  const message = pass
    ? passMessage(styleName, styleValue, actualStyleValue)
    : failMessage(styleName, styleValue, actualStyleValue);

  return { pass, message };
}

export { toHaveStyle };
