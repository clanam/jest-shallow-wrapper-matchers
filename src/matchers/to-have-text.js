import { ShallowWrapper } from 'enzyme';
import { matcherHint, printReceived } from 'jest-matcher-utils';

import { getTextContent } from '../helpers/get-text-content';

function passMessage(expected, actual) {
  return () =>
    matcherHint('.not.toHaveText', 'received', '') +
    `\n\nExpected element not to have text ${expected}:\n` +
    `  ${printReceived(actual)}`;
}

function failMessage(expected, actual) {
  return () =>
    matcherHint('.toHaveText', 'received', '') +
    `\n\nExpected element to have text ${expected}:\n` +
    `  ${printReceived(actual)}`;
}

/** Checks if a shallow-rendered enzyme element has a given text string. */
function toHaveText(receivedElem, text) {
  if (!(receivedElem instanceof ShallowWrapper)) {
    throw new Error(`Expected ${receivedElem} to be an enzyme ShallowWrapper!`);
  }

  const expectedText = Array.isArray(text) ? text : [text];
  const actualText = getTextContent(receivedElem);
  const pass =
    expectedText.length === actualText.length && expectedText.every((t, i) => actualText[i] === t);

  const msgFxn = pass ? passMessage : failMessage;
  const printableActual = actualText.length === 1 ? actualText[0] : actualText;
  const message = msgFxn(JSON.stringify(text), printableActual);

  return { pass, message };
}

export { toHaveText };
