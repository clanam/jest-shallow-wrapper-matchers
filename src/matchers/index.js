import { toBeOfComponentType } from './to-be-of-component-type';
import { toHaveStyle } from './to-have-style';
import { toHaveText } from './to-have-text';

const matchers = {
  toBeOfComponentType,
  toHaveStyle,
  toHaveText,
};

export { matchers };
