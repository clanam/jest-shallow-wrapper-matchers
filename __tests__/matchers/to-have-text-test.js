import React from 'react';
import { Text, View } from 'react-native';
import { shallow } from 'enzyme';

import { toHaveText } from '../../src/matchers/to-have-text';

expect.extend({ toHaveText });

let wrapper;

describe('toHaveText', () => {
  describe('text element', () => {
    beforeEach(() => {
      wrapper = shallow(<Text>Nom nom mango mousse!</Text>);
    });

    it('passes when matching', () => {
      expect(wrapper).toHaveText('Nom nom mango mousse!');
    });

    it('fails when different', () => {
      expect(() =>
        expect(wrapper).toHaveText('Nom nom chocolate mousse!'),
      ).toThrowErrorMatchingSnapshot();
    });
  });

  describe('nested text element', () => {
    beforeEach(() => {
      wrapper = shallow(
        <View>
          <Text>Nom nom mango mousse!</Text>
        </View>,
      );
    });

    it('passes when matching', () => {
      expect(wrapper).toHaveText('Nom nom mango mousse!');
    });

    it('fails when different', () => {
      expect(() =>
        expect(wrapper).toHaveText('Nom nom chocolate mousse!'),
      ).toThrowErrorMatchingSnapshot();
    });
  });

  describe('multiple nested text elements', () => {
    beforeEach(() => {
      wrapper = shallow(
        <View>
          <View>
            <Text>Nom nom mango mousse!</Text>{' '}
          </View>
          <Text>Nom nom nom nom...</Text>
        </View>,
      );
    });

    it('passes when matching', () => {
      expect(wrapper).toHaveText(['Nom nom mango mousse!', 'Nom nom nom nom...']);
    });

    it('fails when different (check single string expectation)', () => {
      expect(() =>
        expect(wrapper).toHaveText('Nom nom chocolate mousse!'),
      ).toThrowErrorMatchingSnapshot();
    });

    it('fails when different (check array expectation, right size)', () => {
      expect(() =>
        expect(wrapper).toHaveText(['Nom nom chocolate mousse!', 'alphabet soup']),
      ).toThrowErrorMatchingSnapshot();
    });
  });

  it('fails when different (check array expectation, too long)', () => {
    expect(() =>
      expect(wrapper).toHaveText([
        'Nom nom mango mousse!',
        'Nom nom nom nom...',
        'eleven squirrels',
      ]),
    ).toThrowErrorMatchingSnapshot();
  });

  it('fails when different (check array expectation, too short)', () => {
    expect(() =>
      expect(wrapper).toHaveText(['Nom nom mango mousse!']),
    ).toThrowErrorMatchingSnapshot();
  });
});

describe('not.toHaveText', () => {
  describe('empty text element', () => {
    beforeEach(() => {
      wrapper = shallow(<Text />);
    });

    it('passes when different', () => {
      expect(wrapper).not.toHaveText('Nom nom mango mousse!');
    });

    it('fails when same', () => {
      expect(() => expect(wrapper).not.toHaveText('')).toThrowErrorMatchingSnapshot();
    });
  });

  describe('text element', () => {
    beforeEach(() => {
      wrapper = shallow(<Text>Nom nom mango mousse!</Text>);
    });

    it('passes when different', () => {
      expect(wrapper).not.toHaveText('Nom nom strawberry mousse!');
    });

    it('fails when same', () => {
      expect(() =>
        expect(wrapper).not.toHaveText('Nom nom mango mousse!'),
      ).toThrowErrorMatchingSnapshot();
    });
  });

  describe('nested text element', () => {
    beforeEach(() => {
      wrapper = shallow(
        <View>
          <Text>Nom nom mango mousse!</Text>
        </View>,
      );
    });

    it('passes when different', () => {
      expect(wrapper).not.toHaveText('Nom nom cherry mousse!');
    });

    it('fails when different', () => {
      expect(() =>
        expect(wrapper).not.toHaveText('Nom nom mango mousse!'),
      ).toThrowErrorMatchingSnapshot();
    });
  });

  describe('multiple nested text elements', () => {
    beforeEach(() => {
      wrapper = shallow(
        <View>
          <View>
            <Text>Nom nom mango mousse!</Text>{' '}
          </View>
          <Text>Nom nom nom nom...</Text>
        </View>,
      );
    });

    it('passes when different', () => {
      expect(wrapper).not.toHaveText(['Nom nom mango mousse!']);
    });

    it('fails when same', () => {
      expect(() =>
        expect(wrapper).not.toHaveText(['Nom nom mango mousse!', 'Nom nom nom nom...']),
      ).toThrowErrorMatchingSnapshot();
    });
  });
});
