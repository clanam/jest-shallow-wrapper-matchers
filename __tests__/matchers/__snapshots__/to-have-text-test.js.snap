// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`not.toHaveText empty text element fails when same 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toHaveText()[22m

Expected element not to have text \\"\\":
  [31m\\"\\"[39m"
`;

exports[`not.toHaveText multiple nested text elements fails when same 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toHaveText()[22m

Expected element not to have text [\\"Nom nom mango mousse!\\",\\"Nom nom nom nom...\\"]:
  [31m[\\"Nom nom mango mousse!\\", \\"Nom nom nom nom...\\"][39m"
`;

exports[`not.toHaveText nested text element fails when different 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toHaveText()[22m

Expected element not to have text \\"Nom nom mango mousse!\\":
  [31m\\"Nom nom mango mousse!\\"[39m"
`;

exports[`not.toHaveText text element fails when same 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toHaveText()[22m

Expected element not to have text \\"Nom nom mango mousse!\\":
  [31m\\"Nom nom mango mousse!\\"[39m"
`;

exports[`toHaveText fails when different (check array expectation, too long) 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toHaveText()[22m

Expected element to have text [\\"Nom nom mango mousse!\\",\\"Nom nom nom nom...\\",\\"eleven squirrels\\"]:
  [31m[\\"Nom nom mango mousse!\\", \\"Nom nom nom nom...\\"][39m"
`;

exports[`toHaveText fails when different (check array expectation, too short) 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toHaveText()[22m

Expected element to have text [\\"Nom nom mango mousse!\\"]:
  [31m[\\"Nom nom mango mousse!\\", \\"Nom nom nom nom...\\"][39m"
`;

exports[`toHaveText multiple nested text elements fails when different (check array expectation, right size) 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toHaveText()[22m

Expected element to have text [\\"Nom nom chocolate mousse!\\",\\"alphabet soup\\"]:
  [31m[\\"Nom nom mango mousse!\\", \\"Nom nom nom nom...\\"][39m"
`;

exports[`toHaveText multiple nested text elements fails when different (check single string expectation) 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toHaveText()[22m

Expected element to have text \\"Nom nom chocolate mousse!\\":
  [31m[\\"Nom nom mango mousse!\\", \\"Nom nom nom nom...\\"][39m"
`;

exports[`toHaveText nested text element fails when different 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toHaveText()[22m

Expected element to have text \\"Nom nom chocolate mousse!\\":
  [31m\\"Nom nom mango mousse!\\"[39m"
`;

exports[`toHaveText text element fails when different 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toHaveText()[22m

Expected element to have text \\"Nom nom chocolate mousse!\\":
  [31m\\"Nom nom mango mousse!\\"[39m"
`;
